core/ocb2progressive.js

```diff
@@ -0,0 +1,137 @@
+/**
+ * OCB2.0 implementation slightly modified by Yifan Gu
+ * to support progressive encryption
+ * @author Yifan Gu
+ */
+
+/** @fileOverview OCB 2.0 implementation
+ *
+ * @author Emily Stark
+ * @author Mike Hamburg
+ * @author Dan Boneh
+ */
+
+/** @namespace
+ * Phil Rogaway's Offset CodeBook mode, version 2.0.
+ * May be covered by US and international patents.
+ *
+ * @author Emily Stark
+ * @author Mike Hamburg
+ * @author Dan Boneh
+ */
+
+sjcl.mode.ocb2progressive = {
+  createEncryptor: function(prp, iv, adata, tlen, premac) {
+    if (sjcl.bitArray.bitLength(iv) !== 128) {
+      throw new sjcl.exception.invalid("ocb iv must be 128 bits");
+    }
+    var i,
+        times2 = sjcl.mode.ocb2._times2,
+        w = sjcl.bitArray,
+        xor = w._xor4,
+        checksum = [0,0,0,0],
+        delta = times2(prp.encrypt(iv)),
+        bi, bl,
+        datacache = [],
+        pad;
+
+    adata = adata || [];
+    tlen  = tlen || 64;
+
+    return {
+      process: function(data){
+        var datalen = sjcl.bitArray.bitLength(data);
+        if (datalen == 0){ // empty input natrually gives empty output
+          return [];
+        }
+        var output = [];
+        datacache = datacache.concat(data);
+        for (i=0; i+4 < datacache.length; i+=4) {
+          /* Encrypt a non-final block */
+          bi = datacache.slice(i,i+4);
+          checksum = xor(checksum, bi);
+          output = output.concat(xor(delta,prp.encrypt(xor(delta, bi))));
```

Comment:
Looking at the decrypt / encrypt function (from the original ocb2 module), it is true that the two looks exceedingly similar (almost seems that encryption is self-inverting) but the main difference, is in the for loops. 

The inverse of encryption is not in the same sequence, making the two not identical, (so OCB2^{-1} \noteq OCB2, OCB2: X -> X ).

After the for loop, the encrypting function adds padding, calculates checksum and append it; the decrypting function, on the other hand, checks it and removes padding. So these two are not the same.

Before the for loop, the two functions are indeed the same, but factoring out that, in my opinion, is not worthwhile.